### DISCUSSION TITLE
Application authentication solution

### DISCUSSION THREAD

**Discussion by helander** [pocketbase/pocketbase#3088] [Q&A] [answered]
I need to come up with a working solution for how to implement authentication for my pocketbase based "App".

The structure of the "App" is something like this:
   - the pocketbase server has a reverse proxy (traefik) as frontend
   - it uses a couple of base collections
   - it has a set of web pages (under public)
   - the web pages loads javascript modules with javascript code that uses the Pocketbase JS SDK

I tested with having the traefik frontend manage basic authentication for the App, but that (I should have known) made the calls from the pages' JS code to the server require the frontend username/password (in addition to the pocketbase users'/admins' credentials).

Any ideas on what the solution architecture could/should look like are most welcome.

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the desired auth flow.

Do you want to combine Basic auth with the PocketBase authentication or just want to use Basic auth to grant permissions to access admin-only collections?

**Reply by helander** [reply]
I am not looking for a particular auth method at this point, but to find a working solution. The solution should be "single-sign-on" in nature, meaning from the web user's point of view you only enter authentication information once per session. 

Does that make it more clear?

**Reply by ganigeorgiev** [reply]
> ... from the web user's point of view you only enter authentication information once per session.

There are no sessions in PocketBase, but `pb.authStore` when running in browser env emulates to some extend a session-like behavior because it uses localStorage for storing the last generated auth token and will be automatically loaded on the next JS SDK initialization on page refresh. Or in other words - you should be able to use just the PocketBase auth in your client-side scripts without the need of the Basic auth of your proxy.

**Reply by helander** [reply]
Just to check that I am following you:
  If I create a login page, where I can enter username and password, and when entered "the page" will call a JS function that will issue an authentication call to the pocketbase server. Then subsequent requests to PAGES in the SAME WINDOW, will automatically when the pocketbase JS module is imported, make sure to reload the auth token from localStorage, so that no more explicit authentication call is required. Correct ? 
  Sounds like a simple solution :)

**Reply by ganigeorgiev** [reply, marked as answer]
Yes, that's correct. If running inside browser env, when calling `new PocketBase(url)` the default AuthStore will try to read the data from the localStorage (if any).

Eventually if you want to ensure that the `pb.authStore.model` has up-to-date user data (eg. in case the user name has changed after the first authentication) you can call `pb.authStore.isValid && await pb.collection("users").authRefresh()` on each page load (it will also ensures that the token `exp` claim is renewed).

**Reply by helander** [reply]
Thanks

**Reply by helander** [reply]
I thought that pb.authStore.clear() would basically invalidate the currently authenticated user, but it seems to not have any such effect. I can fetch data from collections after this call !!!?

**Reply by ganigeorgiev** [reply]
`pb.authStore.clear()` removes the token from the `pb.authStore` instance and localStorage. It doesn't invalidate the token. We currently don't have such method in the web APIs. Tokens are automatically invalidated when resetting the shared app settings secret key or the user unique `tokenKey` (eg. in case of a password reset).


> I can fetch data from collections after this call !!!?

Are you sure that the collection is auth protected? Aka. it has non-empty collection API rules?
You can also try to inspect the browser network request and see if `Authorization: TOKEN` header is being sent with the request (if you call clear() the store authStore be empty and the requests shouldn't have `Authorization` header with them).

**Reply by helander** [reply]
Ok, the collection has empty API rules. I thought that you were not able to access any collection without having been authenticated. 
So this means that it is possible to create a system where you do not require any authentication at all ?

**Reply by ganigeorgiev** [reply]
Yes, you can have mixed access controls. Only if the collection was "locked" will prevent non-admin users or guests access.
Also note that API rules act as server-side filters, aka. for example you can allow returning only records with a specific field value (eg. `status = 'public'`).
If you haven't already, for more details about the API rules you can explore https://pocketbase.io/docs/api-rules-and-filters/.

**Reply by helander** [reply]
OK, thanks

**Reply by helander** [reply]
I have made some experiments and I think I have concluded that I need a solution where I can tell the server to "forget/invalidate" the token of the current user (admin or "normal" user). I.e. independently on if the client(s) continue to send a previously valid token, it would not be accepted by the server. Is this possible to do using the settings API (as you indicated above), if so could you provide some details on how to do it? If not possible via the settings API, could I add a route (routeAdd) that could call some local server function to invalidate the user token?

**Reply by ganigeorgiev** [reply]
If you want to invalidate/reset the token for a single auth record or admin you can call `refreshTokenKey()` on the model and save it. I haven't tested it, but here is a code sample with a custom `POST /invalidate-user-tokens/:id` route accessible only by admins or the current authenticated user:

```js
routerAdd("post", "/invalidate-user-tokens/:id", (c) => {
    const id = c.pathParam("id")

    const record = $app.dao().findRecordById("users", id)

    // reset the unique user token secret
    // (all previously issued record tokens, including the current auth record one, will be invalidated)
    record.refreshTokenKey();
    $app.dao().saveRecord(record)

    return c.noContent(204)
}, $apis.activityLogger($app), $apis.requireAdminOrOwnerAuth("id"))
```

_Note that changing the secrets through the app settings web API will invalidate the tokens for all users._

**Reply by helander** [reply]
Thanks, I will try this.
Regarding that settings web API will invalidate tokens for all users is an important piece of information .

**Reply by helander** [reply]
Just by reading the code, I do not understand how it could work for a logged in admin user , it seems only to operate on users from the 'users' collection !?

**Reply by ganigeorgiev** [reply]
`refreshTokenKey()` is also available for the admin models, aka.:

```js
const admin = $app.dao().findAdminById(id)

admin.refreshTokenKey();

$app.dao().saveAdmin(admin);
```

**Reply by helander** [reply]
So far I can not get the calls pass the middleware that requires admin or owner authentication, so I just removed it for now (I should read up on this functionality and see if I can get that to work). Meanwhile I tested with this middleware removed and then it seems to work fine to refresh admin tokens (after that operation I can no longer call services that require admin privileges). For "owner" auth, I can still call and get result from the 'users' collection, after having refreshed the "owner" token. 
Any ideas?

**Reply by ganigeorgiev** [reply]
Sorry, I'm confused. I don't have the full picture of your use case and I have no idea what you are trying to do and why do you want to invalidate the admins and user tokens manually in the first place.
Keep in mind that admins and auth records are 2 different entities and usually on application level you interact mostly with the latter.

**Reply by helander** [reply]
My main issue it is that in order to "logout" a browser application, you need to make sure that the tokens the client passes to the server no longer are valid. You could possibly do that in a client, but it requires you to send event to all pages of the applications and make sure that each pages cleans up any token information it has cached. You will then require the code on each page to do this, and this is not a very 'reliable' solution.
So a better approach is if code on any page would signal to the server that the token that is shared among the pages no longer is valid, and that the server no longer will accept calls using this token.

**Reply by ganigeorgiev** [reply]
You can use listen to the [`storage` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event) to sync the auth state across multiple pages (_I'll consider having it builtin for the default `LocalAuthStore` with the ongoing SDK refactoring_). Something like (_I haven't tested it_):

```js
// something like this will be done automatically by the sdk in the next release
window.addEventListener("storage", () => {
  if (!pb.authStore.token) { // this getter retrieves the token from the auth store
    pb.authStore.clear();
  }
});
```

Note that refreshing the token key invalidates all tokens for that user/admin, not just the current one. We don't have mechanism to invalidate only a single token because they are not stored on the server. We can only invalidate all tokens for all users/admins OR all tokens for a single individual user/admin.

I don't know why the record token invalidation doesn't work for you. I suspect that there is something wrong with your collection API rule(s) (eg. missing or invalid auth check).

**Reply by helander** [reply]
An additional question/comment:

The documentation mentions three different types of users - admins, authorized users and guests. 
I guess that guests are users that has not issued any authentication calls towards the server, correct?
For such users, there will be no auth records in the requests to the server, correct?
This would mean for this kind of user there is no way to have different authotization rules depending on who the user is, because there is no info in the request about who the user is, correct?

Since authorization is only supported by the collections and not per user, I suggest to use the term ’authenticated user’ instead of ’authorized user’.

**Reply by ganigeorgiev** [reply]
> I guess that guests are users that has not issued any authentication calls towards the server, correct?
For such users, there will be no auth records in the requests to the server, correct?
This would mean for this kind of user there is no way to have different authotization rules depending on who the user is, because there is no info in the request about who the user is, correct?

Yes, that's correct. If you want to implement an "anonymous user", eg. to store some data of the guest interaction so that later they can continue from where they left, you can create a regular user record with random password that could be stored locally. Once the user decide to "officially create" their account you can show the user a screen to enter their own new password and behind the scene you can send an `update()` call with that new password.

**Reply by helander** [reply]
In order to understand how authentication and authorization works in pocketbase I created a test program that in turn authenticates as an admin and then two different users from the 'user' collection. After authentication I call a test that reads from different collections:

- one that is locked
- one with an empty rule
-  one with the rule `request.auth.id !=""`

Before calling these tests as authenticated users I call the same tests as 'anonymous', i.e. no authentication performed prior to calling the tests.

What puzzled me was that the anonymous 'user' where allowed to fetch the records from the collection with the rule `request.auth.id !=""`. Since this happened I assume that my expectations were wrong. Could you explain how to disable anonymous users from accessing a collection while authenticated users are able to access the same collection?

**Reply by ganigeorgiev** [reply]
In general if the request was sent without being authenticated (aka. without `Authorization: TOKEN` header) then the rule `@request.auth.id != ""` should prevent access to the collection record(s).

If by "anonymous user" you mean the same as in https://github.com/pocketbase/pocketbase/discussions/3088#discussioncomment-6698546, then the request is technically authorized in this case. You can add a helper `bool` field in the `users` collection that could indicate "anonymous user" and check it in the rule like `@request.auth.id != "" && @request.auth.isAnonymous = false`.

If that's not what you are looking for, please consider opening a new disscusion with more details about your collections structure and what you are trying to do and what is the desired outcome.

**Reply by helander** [reply]
Befor opening a new thread I just want to make it clear that by anonymous I mean that no authentication call has been made at all by the client prior to calling the collection access functions, but the rule does not seem to stop such requests. Starting a new thread I guess I shall attach my JS code and also the setup of the collection in question. Is there some way to dump the collection information (including rules) so that I can attach to the discussion?

**Reply by ganigeorgiev** [reply]
> I mean that no authentication call has been made at all by the client prior to calling the collection access functions, but the rule does not seem to stop such requests.

No, unless it is a bug (but we have several integration tests that should cover this use case) it shouldn't behave this way.

Double check the request payload to confirm that there is no `Authorization: TOKEN` header sent with the request. Keep in mind that the default JS SDKs AuthStore uses localStorage and will automatically load the previously stored local authstate on JS SDK instantiation.

> Is there some way to dump the collection information (including rules) so that I can attach to the discussion?

You can export your collections configuration as JSON from the _Admin UI > Settings > Export collections_.

**Reply by helander** [reply]
Thanks.
I am not running these tests in a browser, but running in deno on a linux host. I do not know if localStorage is available there, I have to check prior to creating a thread.

**Reply by helander** [reply]
Clearing the pocketbase info from localStorage made it work as expected, thanks. No new thread will be created.

**Reply by ganigeorgiev** [reply]
I don't have much experience with Deno but indeed it seems that `LocalStorage` is supported (at least based on the [MDN compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage#browser_compatibility)).

I'll consider adding a note in the JS SDK as it could be a security issue for those who use Deno since I guess the localStorage will be shared by all clients (_edit: the rules for the Deno localStorage are a little more complicated - https://deno.land/manual@v1.36.1/runtime/web_storage_api#web-storage-api_).

**Reply by helander** [reply]
Is there a similar way to export the collection records, not just the collection metadata ?

**Reply by ganigeorgiev** [reply]
No currently we don't have an export records functionality (it will be eventually implemented as part of #48).
